pkg/adaptor/rethinkdb/reader.go

```diff
@@ -0,0 +1,197 @@
+package rethinkdb
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/ops"
+
+	re "gopkg.in/gorethink/gorethink.v3"
+)
+
+var (
+	_ client.Reader = &Reader{}
+)
+
+// Reader fulfills the client.Reader interface for use with both copying and tailing a RethinkDB
+// database.
+type Reader struct {
+	db   string
+	tail bool
+}
+
+func newReader(db string, tail bool) client.Reader {
+	return &Reader{db, tail}
+}
+
+type iterationComplete struct {
+	cursor *re.Cursor
+	table  string
+}
+
+func (r *Reader) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
+	return func(s client.Session, done chan struct{}) (chan message.Msg, error) {
+		out := make(chan message.Msg)
+		session := s.(*Session).session
+		go func() {
+			defer close(out)
+			log.With("db", r.db).Infoln("starting Read func")
+			tables, err := r.listTables(session, filterFn)
+			if err != nil {
+				log.With("db", r.db).Errorf("unable to list tables, %s", err)
+				return
+			}
+			iterationComplete := r.iterateTable(session, tables, out, done)
+			var wg sync.WaitGroup
+			func() {
+				for {
+					select {
+					case <-done:
+						return
+					case i, ok := <-iterationComplete:
+						if !ok {
+							return
+						}
+						log.With("db", r.db).With("table", i.table).Infoln("iterating complete")
+						if i.cursor != nil {
+							go func(wg *sync.WaitGroup, t string, c *re.Cursor) {
+								wg.Add(1)
+								defer wg.Done()
+								errc := r.sendChanges(t, c, out, done)
+								for err := range errc {
+									log.With("db", r.db).With("table", t).Errorln(err)
+									return
+								}
+							}(&wg, i.table, i.cursor)
+						}
+					}
+				}
+			}()
+			log.With("db", r.db).Infoln("Read completed")
+			go func() {
```

Comment:
yea... I had some issues with trying to still receive on the done channel when doing a tail but it's very likely that the wait group accomplishes the same thing since it gets propagated down to `sendChanges`. I'll remove it and see what happens.